### PR TITLE
A few edits to fix ClassLoader leaks

### PR DESCRIPTION
**Pull request by greenjustin**
Motivation:

We have a need to shut down and start up a server built on Netty with 0 ClassLoader leaks.

Modification:

Most of these changes are to prevent old ClassLoaders from being trapped in the backtrace variable of Exceptions. The stack traces of these Exceptions are empty, but there is another variable in Throwable called backtrace that is somehow populated from JVM native code. Because there doesn't seem to be an easy way to get rid of this, the solution we came up with was creating dummy instances of e.g. HpackDecoder from a class loaded using the System ClassLoader so the backtrace won't be tainted with a custom ClassLoader. In order for this to work, we have to tweak a few visibilities.

The change to GlobalEventExecutor just allows us to drain the execution queue on server shutdown.

### CODE REVIEW

File: common/src/main/java/io/netty/util/concurrent/GlobalEventExecutor.java [hunk 1]
```diff
@@ -159,7 +159,15 @@ public boolean inEventLoop(Thread thread) {
     @Override
     @Deprecated
     public void shutdown() {
-        throw new UnsupportedOperationException();
```
Comment (hyperxpro) [inline common/src/main/java/io/netty/util/concurrent/GlobalEventExecutor.java:162]: API breakage

File: codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java [hunk 2]
```diff
@@ -51,7 +51,7 @@
 import static io.netty.util.internal.ObjectUtil.checkPositive;
 import static io.netty.util.internal.ThrowableUtil.unknownStackTrace;
 
-final class HpackDecoder {
+public final class HpackDecoder {
```
Comment (hyperxpro) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java:54]: Why make this `public`?

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java [hunk 3]
```diff
@@ -390,8 +390,8 @@ protected void run() {
                     //increase the size of the array as we needed the whole space for the events
                     events.increase();
                 }
-            } catch (Throwable t) {
```
Comment (hyperxpro) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java:393]: Use `throwable`.

**Review by hyperxpro — CHANGES_REQUESTED** [review]
Lots of API change and breakage.

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java [hunk 4]
```diff
@@ -390,8 +390,8 @@ protected void run() {
...
-                handleLoopException(t);
+            } catch (Exception e) {
+                handleLoopException(e);
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java:394]: why this change ? Thins looks not right...

File: common/src/main/java/io/netty/util/concurrent/GlobalEventExecutor.java [hunk 5]
```diff
@@ -159,7 +159,15 @@ public boolean inEventLoop(Thread thread) {
...
+        taskQueue = new LinkedBlockingQueue<Runnable>();
+        quietPeriodTask = new ScheduledFutureTask<Void>(
+           this, Executors.<Void>callable(new Runnable() {
+               @Override
+               public void run() {
+                // NOOP
+               }
+           }, null), ScheduledFutureTask.deadlineNanos(SCHEDULE_QUIET_PERIOD_INTERVAL),
+           -SCHEDULE_QUIET_PERIOD_INTERVAL);
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/concurrent/GlobalEventExecutor.java:170]: What you are doing here ? This is also no thread-safe at all.

File: common/src/main/java/io/netty/util/concurrent/DefaultPromise.java [hunk 6]
```diff
@@ -86,7 +86,7 @@ public DefaultPromise(EventExecutor executor) {
     /**
      * See {@link #executor()} for expectations of the executor.
      */
-    protected DefaultPromise() {
+    public DefaultPromise() {
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/concurrent/DefaultPromise.java:89]: Honestly I think we need to fix the in another way...  I am not comfortable with changing the access levels and so make it "public api" (this comment is true for all the changes)

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: common/src/main/java/io/netty/util/concurrent/GlobalEventExecutor.java (hunk 1 above)
Comment (normanmaurer) [reply common/src/main/java/io/netty/util/concurrent/GlobalEventExecutor.java:162]: nope thats not a breakage...

**Comment by normanmaurer** [comment]
I think I have an idea how to solve this... Gimme a few minutes

**Comment by normanmaurer** [comment]
@greenjustin can you test https://github.com/netty/netty/pull/10691 ?

**Comment by normanmaurer** [comment]
Also for the `GlobalEventExecutor` case you may just want to use `awaitInactivity(...)` to ensure the task queue is empty

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java (hunk 4 above)
Comment (greenjustin) [reply transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java:394]: Sorry, I should have mentioned this in the original comment. We also have a need to be able to kill threads using Thread.stop(). Under the hood, Thread.stop actually kills threads using ThreadDeathError. Unfortunately, this means that if you swallow all Throwables, the thread becomes unkillable.

**Comment by greenjustin** [comment]
> @greenjustin can you test #10691 ?

Sure, give me a few minutes to check. Thank you so much for helping with this!

**Comment by greenjustin** [comment]
> @greenjustin can you test #10691 ?

This patch works well for removing the backtrace problem. Thanks!

**Comment by greenjustin** [comment]
> Also for the `GlobalEventExecutor` case you may just want to use `awaitInactivity(...)` to ensure the task queue is empty

I'm not sure this really works for our use case. This will block until the execution queue is drained, right? But we actually want to forcibly cancel all jobs in the queue.

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java (hunk 4 above)
Comment (normanmaurer) [reply transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java:394]: Why you just cant normally shutdown the `EventLoop` ? Like `EventLoopGroup.shutdownGracefully().sync()` ?

**Comment by normanmaurer** [comment]
@greenjustin the question is what does cancel here really mean ? As some of the things there may be related to cleanup etc so they should run.

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java (hunk 4 above)
Comment (greenjustin) [reply transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java:394]: This is for infrastructure code that doesn't necessarily have knowledge about when EventLoops are created. Our clients are free to create them as they please and we need a generic, scalable solution for taking them down as quickly as possible.

Comment (normanmaurer) [reply transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java:394]: Hmm ... Thread.stop() was removed some time ago . How will think work for you with newer JDK versions ?

**Comment by greenjustin** [comment]
> @greenjustin the question is what does cancel here really mean ? As some of the things there may be related to cleanup etc so they should run.

We probably don't need any of these? I'm not sure, what specific cleanup operations are performed?

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java (hunk 4 above)
Comment (greenjustin) [reply transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java:394]: Was it removed? In the version we use, it is still just deprecated. We might be petitioning the OpenJVM maintainers to keep it around or forking the JVM internally just for that purpose.

File: resolver/src/main/java/io/netty/resolver/AddressResolverGroup.java [hunk 7]
```diff
@@ -117,7 +117,9 @@ public void close() {
         }
 
         for (final Map.Entry<EventExecutor, GenericFutureListener<Future<Object>>> entry : listeners) {
-            entry.getKey().terminationFuture().removeListener(entry.getValue());
+            if (entry.getKey().terminationFuture() != null) {
+                entry.getKey().terminationFuture().removeListener(entry.getValue());
```
Comment (normanmaurer) [inline resolver/src/main/java/io/netty/resolver/AddressResolverGroup.java:121]: why this change ?

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java (hunk 4 above)
Comment (normanmaurer) [reply transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java:394]: Yes in JDK11: https://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8204243

File: resolver/src/main/java/io/netty/resolver/AddressResolverGroup.java (hunk 7 above)
Comment (greenjustin) [reply resolver/src/main/java/io/netty/resolver/AddressResolverGroup.java:121]: We get null pointer exceptions otherwise. I'm not entirely sure why, it might have to do with our use of Thread.stop().

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java (hunk 4 above)
Comment (greenjustin) [reply transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java:394]: Oh, this is for the official JDK though, right? We are using OpenJDK11.

In either case, I see that our motivation for this change is rather unique to our circumstances. Is there a specific reason to swallow all Throwables here? Errors are supposed to be so severe that no application could reasonably handle them (like OutOfMemory, JVM errors, etc).

Comment (normanmaurer) [reply transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java:394]: As far as I know it’s the same for OpenJDK... I guess we could rethrow the ThreadDeath tho

Comment (greenjustin) [reply transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java]: That works for me. How does this most recent commit look?

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java (hunk 3 above)
Comment (normanmaurer) [reply transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java:393]: Can we have this change as a separate PR but for all EventLoop implementations ?

Comment (greenjustin) [reply transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java:393]: Sure, created https://github.com/netty/netty/pull/10694

File: common/src/main/java/io/netty/util/concurrent/GlobalEventExecutor.java (hunk 5 above)
Comment (normanmaurer) [reply common/src/main/java/io/netty/util/concurrent/GlobalEventExecutor.java]: I think this will not be good enough... let me think about it

File: resolver/src/main/java/io/netty/resolver/AddressResolverGroup.java (hunk 7 above)
Comment (normanmaurer) [reply resolver/src/main/java/io/netty/resolver/AddressResolverGroup.java:121]: hmm... this should never happen.

**Comment by greenjustin** [comment]
Upon further testing, it looks like the GlobalEventExecutor.java and AddressResolverGroup.java changes are no longer necessary, the ThreadDeath rethrow fix and backtrace leak fix are sufficient.
